### DISCUSSION TITLE
Map renderWorldCopies regression

### DISCUSSION THREAD

**Discussion by raboczi** [maplibre/maplibre-gl-js#5024] [General]
Starting with v5.0.0-pre.6, a Mercator projection no longer renders world copies and clamps panning and zooming to prevent entering the unrendered area. It's consequently no longer possible to view an area which crosses the antimeridian, except in the Globe projection. This used to work as recently as v5.0.0-pre.5.

**Comment by HarelM** [comment]
@NathanMOlson @kubapelc can you take a look at this issue and see if anything you did broke this? 
@raboczi can you create a minimal reproduction jsbin/codepen?

**Reply by raboczi** [reply]
It seems like if you start in Mercator it's okay, but if you switch the projection from Globe and back to Mercator, that's when things break.  Here's the best I can do for a test case at the moment:

1. Unpack dist.zip for v5.0.0-pre.6
2. Add the following index.html alongside it
  ```
  <html lang="en">
    <head>
      <link rel="stylesheet" href="dist/maplibre-gl.css"/>
      <script src="dist/maplibre-gl.js"></script>
    </head>
    <body>
      <div id="container" style="height:100%; width:100%"></div>
      <script>
        const MAPTILER_KEY = 'get_your_own_OpIi9ZULNHzrESv6T2vL';
        var map = new maplibregl.Map({
          container: 'container',
          style: `https://api.maptiler.com/maps/basic-v2/style.json?key=${MAPTILER_KEY}`,
          zoom: 5
        })
        map.on('style.load', () => {
          map.setProjection({ type: 'globe' })
          map.setCenter([179, 65])
          map.setProjection({ type: 'mercator' })
        })
      </script>
    </body>
  </html>
  ```
3. Serve the directory, e.g. with `npx http-server . --port 80`
4. Browse http://localhost

Only half the map will appear.  The moment you manually zoom or pan, you'll be shunted away from the antimeridian and unable to cross it.  Happens consistently under Firefox, Opera, and Safari on MacOS 15.

**Reply by NathanMOlson** [reply]
I'm guessing it's this change in `GlobeTransform`:

```
    setRenderWorldCopies(_renderWorldCopies: boolean): void {
    }
```

```
    get renderWorldCopies(): boolean {
        return false;
    }
```

This was done to unify this code in `coveringTiles()`:

```
    if (transform.renderWorldCopies) {
        // Render copy of the globe thrice on both sides
        for (let i = 1; i <= 3; i++) {
            stack.push(newRootTile(-i));
            stack.push(newRootTile(i));
        }
    }
```

I'll put in a fix for this.

**Comment by zerda-ocm** [comment]
Because of the mention of the antimeridian, possibly related to this: https://github.com/maplibre/maplibre-gl-js/issues/5009

**Comment by NathanMOlson** [comment]
Hopefully this will fix it. I don't have time to try it out at the moment.

https://github.com/maplibre/maplibre-gl-js/pull/5027

**Comment by raboczi** [comment]
I tinkered a bit and tried the following:
```
console.log('A', map.getRenderWorldCopies())
map.setProjection({ type: 'globe' })
console.log('B', map.getRenderWorldCopies())
map.setCenter([179, 65])
map.setProjection({ type: 'mercator' })
console.log('C', map.getRenderWorldCopies())
map.setRenderWorldCopies(true)
```
This workaround results in the map being displayed as expected, and the following log:
```
A - true
B - false
C - false
```
So it looks like it's just that setting the projection to Globe has the side effect of clearing the renderWorldCopies flag.

**Comment by NathanMOlson** [comment]
This is now fixed by #5027, although I don't love the additional complexity.
